src/main/java/randoop/generation/DemandDrivenInputCreator.java

```diff
@@ -0,0 +1,446 @@
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import randoop.DummyVisitor;
+import randoop.ExecutionOutcome;
+import randoop.NormalExecution;
+import randoop.main.RandoopUsageError;
+import randoop.operation.CallableOperation;
+import randoop.operation.ConstructorCall;
+import randoop.operation.MethodCall;
+import randoop.operation.TypedClassOperation;
+import randoop.operation.TypedOperation;
+import randoop.sequence.ExecutableSequence;
+import randoop.sequence.Sequence;
+import randoop.sequence.SequenceCollection;
+import randoop.test.DummyCheckGenerator;
+import randoop.types.ArrayType;
+import randoop.types.NonParameterizedType;
+import randoop.types.Type;
+import randoop.types.TypeTuple;
+import randoop.util.EquivalenceChecker;
+import randoop.util.Log;
+import randoop.util.Randomness;
+import randoop.util.SimpleArrayList;
+import randoop.util.SimpleList;
+
+/**
+ * A demand-driven approach to construct inputs. Randoop works by selecting a method and then trying
+ * to find inputs for that method. Ordinarily, Randoop works bottom-up: if Randoop cannot find
+ * inputs for the selected method, it gives up and selects a different method. This demand-driven
+ * approach works top-down: if Randoop cannot find inputs for the selected method, then it looks for
+ * methods that create values of the necessary type, and iteratively tries to call them.
+ *
+ * <p>The demand-driven approach implements the "Detective" component described by the ASE 2015
+ * paper <a href="https://people.kth.se/~artho/papers/lei-ase2015.pdf">"GRT: Program-Analysis-Guided
+ * Random Testing" by Ma et al.</a> .
+ */
+public class DemandDrivenInputCreator {
+  /**
+   * The principal set of sequences used to create other, larger sequences by the generator. New
+   * sequences are added on demand for creating object of missing types. Shared with {@link
+   * ComponentManager#gralComponents}.
+   */
+  private final SequenceCollection sequenceCollection;
+
+  /**
+   * If true, {@link #createInputForType(Type)} returns only sequences that declare values of the
+   * exact type that was requested.
+   */
+  private boolean exactTypeMatch;
+
+  /**
+   * If true, {@link #createInputForType(Type)} returns only sequences that are appropriate to use
+   * as a method call receiver, i.e., Type.isNonreceiverType() returns false for the type of the
+   * variable created by the sequence.
+   */
+  private boolean onlyReceivers;
+
+  // TODO: The original paper uses a "secondary object pool (SequenceCollection in Randoop)"
+  // to store the results of the demand-driven input creation. This theoretically reduces
+  // the search space for the missing types. Consider implementing this feature and test whether
+  // it improves the performance.
```

Comment:
We should probably do this if Randoop's pool size explodes due to input construction (chain of objects to consrtruct)